iroh/src/magicsock.rs

```diff
@@ -1561,7 +1633,50 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
-type RelayRecvResult = Result<(PublicKey, quinn_udp::RecvMeta, Bytes), io::Error>;
+/// Channel for [`MagicSock::poll_recv_relay`] to receive QUIC datagrams from relays.
+///
+/// This channel contains both the send and receive sides.  The receive side implements the
+/// poll interface with required wakers.
+#[derive(Debug)]
+struct RelayRecvChannel {
+    sender: mpsc::Sender<RelayReadResult>,
+    receiver: parking_lot::Mutex<mpsc::Receiver<RelayReadResult>>,
```

Comment:
could this just be `Mutex<VecDeque<RelayReadResult>>>`?

Reply:
WRT to all comments on the choice of channel and waker: these did cross my mind.  There are probably also channel implementations that might give us the desired wakers behaviour out of the box (I think both flume and async_channle have stream APIs that are probably usable).

However I explicitly chose to keep these the same: I was targetting this change to the architectural change and didn't feel like also changing the implementation here.

Anyway, if folks would like to see this change at the same time I'm happy to do so.

So, onto this particular suggestion: using a `VecDeque` would make this an unbounded queue I think which is probably a bad idea.  I would expect a fixed-size channel to be more efficiently implemented than a variable-size one.

Reply:
FWIW, if you made all of this `ConcurrentQueue` (from `concurrent-queue`), and keep the waker, then you essentially are 80% towards an implementation of `async-channel` :upside_down_face: 

But no seriously, I think this might be a good case for `ConcurrentQueue` (for an eventual follow-up PR).

Reply:
Postponed to followup, left a todo message for it.